finality-aleph/src/network/address_cache.rs

```diff
@@ -0,0 +1,180 @@
+/// as content of this struct rely on unstable substrate's API and can change in future.
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
+pub struct AdditionalP2PAddressingInfo {
+    pub p2p_network_peer_id: String,
+    pub version_string: Option<String>,
+}
+
+/// Stores most recent information about validator addresses.
+#[derive(Debug, Clone)]
+pub struct ValidatorAddressCache {
+    data: Arc<Mutex<LruCache<AccountId, ValidatorAddressingInfo>>>,
+}
+
+const VALIDATOR_ADDRESS_CACHE_SIZE: usize = 300;
+
+impl ValidatorAddressCache {
+    pub fn new() -> Self {
+        Self {
+            data: Arc::new(Mutex::new(LruCache::new(
+                NonZeroUsize::try_from(VALIDATOR_ADDRESS_CACHE_SIZE).unwrap(),
+            ))),
+        }
+    }
+
+    pub fn insert(&self, validator_stash: AccountId, info: ValidatorAddressingInfo) {
+        self.data.lock().put(validator_stash, info);
+    }
+
+    pub fn read(&self) -> HashMap<AccountId, ValidatorAddressingInfo> {
+        HashMap::from_iter(self.data.lock().iter().map(|(k, v)| (k.clone(), v.clone())))
+    }
+}
+
+impl Default for ValidatorAddressCache {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub trait ValidatorAddressCacheUpdater {
+    /// In session `SessionIndex`, validator `NodeIndex` was using addresses specified in `most_recent_info`.
+    /// A session and validator_index identify the validator uniquely.
+    fn update(
+        &self,
+        session: SessionId,
+        validator_index: NodeIndex,
+        most_recent_info: ValidatorAddressingInfo,
+    );
+}
+
+pub struct ValidatorAddressCacheUpdaterImpl<K: KeyOwnerInfoProvider, A: AuthorityProvider> {
+    validator_address_cache: ValidatorAddressCache,
+    key_owner_info_provider: K,
+    authority_provider: A,
+    session_boundary_info: SessionBoundaryInfo,
+}
+
+impl<K: KeyOwnerInfoProvider, A: AuthorityProvider> ValidatorAddressCacheUpdaterImpl<K, A> {
+    pub fn new(
+        validator_address_cache: ValidatorAddressCache,
+        key_owner_info_provider: K,
+        authority_provider: A,
+        session_boundary_info: SessionBoundaryInfo,
+    ) -> Self {
+        Self {
+            validator_address_cache,
+            key_owner_info_provider,
+            authority_provider,
+            session_boundary_info,
+        }
+    }
+}
+
+impl<K: KeyOwnerInfoProvider, A: AuthorityProvider> ValidatorAddressCacheUpdater
+    for ValidatorAddressCacheUpdaterImpl<K, A>
+{
+    fn update(
+        &self,
+        session: SessionId,
+        validator_index: NodeIndex,
+        info: ValidatorAddressingInfo,
+    ) {
+        let block = self
+            .session_boundary_info
+            .boundaries_for_session(session)
+            .first_block();
+
+        if let Some(authority_data) = self.authority_provider.authority_data(block) {
+            let aleph_key = authority_data.authorities()[validator_index.0].clone();
+            if let Some(validator_stash) = self
```

Comment:
make fn instead that return an owner for a given block and index.
`fn owner_at_block(block, index) -> Option<Owner>`

Reply:
or instead of block use session also

Reply:
Refactored as you suggested in #1449, however, after making updating logic optional and introducing an enum this has both up and downsides :/